File: Aigilas/Aigilas/Strategies/TargetSet.cs
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using Microsoft.Xna.Framework;
using SPX.Entities;
using Aigilas.Creatures;
using Aigilas.Entities;
using SPX.Core;

namespace Aigilas.Strategies
{
    public class TargetSet
    {
        private List<ICreature> _targets = new List<ICreature>();
        private List<int> _targetActorTypes = new List<int>();
        private List<int> _targetEntityTypes = new List<int>();
        private readonly ICreature _parent;

        public TargetSet(ICreature parent)
        {
            _parent = parent;
        }

        public ICreature AddTarget(ICreature target)
        {
            _targets.Add(target);
            return target;
        }

        public void AddTargetTypes(params int[] actorTypes)
        {
            foreach(var type in actorTypes)
            {
                _targetActorTypes.Add(type);
            }
        }

        public void AddEntityTypes(params int[] entityTypes)
        {
            foreach (var type in entityTypes)
            {
                _targetEntityTypes.Add(type);
            }
        }

        public void AddTargets(ICreature source)
        {
            _targets.AddRange(source.GetTargets()._targets);
            _targetActorTypes.AddRange(source.GetTargets()._targetActorTypes);
        }

        public IEnumerable<ICreature> AddTargets(IEnumerable<ICreature> targets)
        {
            if (targets != null)
            {
                _targets.AddRange(targets);
            }
            return targets;
        }

        private List<ICreature> _calculatedTargets;
        private float dist;
        ICreature closest;
        public ICreature FindClosest()
        {
            closest = null;
            var closestDistance = float.MaxValue;
            if (_targets != null)
            {
                foreach (var target in _targets)
                {
                    dist = Point2.DistanceSquared(target.GetLocation(), _parent.GetLocation());
                    if (dist < closestDistance)
                    {
                        closest = target;
                        closestDistance = dist;
                    }
                }
                foreach(var actorType in _targetActorTypes)
                {
                    _calculatedTargets = EntityManager.GetActors(actorType).Select(a=>a as ICreature).ToList();
                    foreach(var creature in _calculatedTargets)
                    {
                        if (creature != _parent)
                        {
                            dist = Point2.DistanceSquared(creature.GetLocation(), _parent.GetLocation());
                            if (dist < closestDistance)
                            {
                                closest = creature;
                                closestDistance = dist;
                            }
                        }
                    }
                }
            }
            return closest;
        }

        public IEntity GetCollidedTarget(Entity source)
        {
            foreach(var target in _targets)
            {
                if(target.Contains(source.GetLocation()))
                {
                    return target;
                }
            }

            foreach (var actorType in _targetActorTypes)
            {
                foreach (var target in EntityManager.GetActorsAt(source.GetLocation()))
                {
                    if (target.GetActorType() == actorType ||
                        (actorType == ActorType.NONPLAYER && target.GetActorType() != ActorType.PLAYER) ||
                        (actorType == ActorType.PLAYER && target.GetActorType() == ActorType.PLAYER))
                    {
                        return target as IEntity;
                    }
                }
            }
            return null;
        }

        internal IEnumerable<int> GetTargetActorTypes()
        {
            return _targetActorTypes;
        }
    }
}

File: aigilas/src/aigilas/strategies/TargetSet.java
package aigilas.strategies;

import aigilas.creatures.BaseCreature;
import sps.bridge.ActorType;
import sps.bridge.ActorTypes;
import sps.core.Core;
import sps.core.Point2;
import sps.entities.Entity;
import sps.entities.EntityManager;
import sps.entities.IActor;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TargetSet

{
    private final List<BaseCreature> _targets = new ArrayList<BaseCreature>();
    private final List<ActorType> _targetActorTypes = new ArrayList<ActorType>();
    private final List<Integer> _targetEntityTypes = new ArrayList<Integer>();
    private final BaseCreature _parent;

    public TargetSet(BaseCreature parent)

    {
        _parent = parent;
    }

    public BaseCreature addTarget(BaseCreature target)

    {
        _targets.add(target);
        return target;
    }

    public void addTargetTypes(ActorType... actorTypes) {
        Collections.addAll(_targetActorTypes, actorTypes);
    }

    public void addEntityTypes(int... entityTypes) {
        for (int type : entityTypes) {
            _targetEntityTypes.add(type);
        }
    }

    public void addTargets(BaseCreature source)

    {
        _targets.addAll(source.getTargets()._targets);
        _targetActorTypes.addAll(source.getTargets()._targetActorTypes);
    }

    public List<BaseCreature> addTargets(List<BaseCreature> targets)

    {
        if (targets != null) {
            _targets.addAll(targets);
        }
        return targets;
    }

    private List<BaseCreature> _calculatedTargets;
    private float dist;
    BaseCreature closest;

    public BaseCreature findClosest() {
        closest = null;
        float closestDistance = Float.MAX_VALUE;
        if (_targets != null) {
            for (BaseCreature target : _targets) {
                dist = Point2.distanceSquared(target.getLocation(), _parent.getLocation());
                if (dist < closestDistance) {
                    closest = target;
                    closestDistance = dist;
                }
            }
            for (ActorType actorType : _targetActorTypes) {
                _calculatedTargets = new ArrayList<BaseCreature>();
                for (IActor entity : EntityManager.get().getActors(actorType)) {
                    _calculatedTargets.add((BaseCreature) entity);
                }

                for (BaseCreature creature : _calculatedTargets) {
                    if (creature != _parent) {
                        dist = Point2.distanceSquared(creature.getLocation(), _parent.getLocation());
                        if (dist < closestDistance) {
                            closest = creature;
                            closestDistance = dist;
                        }
                    }
                }
            }
        }
        return closest;
    }

    public Entity getCollidedTarget(Entity source)

    {
        for (BaseCreature target : _targets) {
            if (target.contains(source.getLocation())) {
                return target;
            }
        }

        for (ActorType actorType : _targetActorTypes) {
            for (IActor target : EntityManager.get().getActorsAt(source.getLocation())) {
                if (target.getActorType() == actorType || (actorType == ActorTypes.get(Core.Non_Player) && target.getActorType() != ActorTypes.get(Core.Player)) || (actorType == ActorTypes.get(Core.Player) && target.getActorType() == ActorTypes.get(Core.Player))) {
                    return (Entity) target;
                }
            }
        }
        return null;
    }

    public List<ActorType> getTargetActorTypes() {
        return _targetActorTypes;
    }
}
